src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,487 @@
+                        # shrink the radius
+                        radius *= shrink_factor
+
+                        msg = 'Updated model output larger than previous output.'
+
+            # handle parameter rejection
+            if not rejected:
+                data['mus'].append(mu.copy())
+                mu_norm = np.linalg.norm(mu)
+                update_norms.append(np.linalg.norm(mu - old_mu))
+
+                data['subproblem_data'].append(sub_data)
+
+                with logger.block('Computing first order criticality...'):
+                    # fom.output_d_mu is potentially for free after enrichment, e.g. using caching
+                    gradient = fom.parameters.parse(fom.output_d_mu(mu)).to_numpy()
+                    first_order_criticality = np.linalg.norm(mu - parameter_space.clip(mu - gradient).to_numpy())
+                    foc_norms.append(first_order_criticality)
+
+                surrogate.accept()
+                logger.info(f'Current mu iterate accepted: {msg}')
+
+                old_rom_output = current_output
+            else:
+                mu = old_mu
+                surrogate.reject()
+                logger.info(f'Current mu iterate rejected: {msg}')
+
+            if not np.isfinite(mu_norm):
+                raise TRError('Failed to converge.')
+
+    logger.info('')
+
+    data['update_norms'] = np.array(update_norms)
+    data['foc_norms'] = np.array(foc_norms)
+    data['iterations'] = iteration
+    data['rom_evaluations'] = surrogate.rom_evaluations
+    data['enrichments'] = surrogate.enrichments
+
+    return mu, data
+
+def coercive_rb_trust_region(reductor, primal_dual=False, parameter_space=None, initial_guess=None, beta=.95, radius=.1,
+                             shrink_factor=.5, miniter=0, maxiter=30, miniter_subproblem=0, maxiter_subproblem=400,
+                             tol_criticality=1e-6, radius_tol=.75, rtol_output=1e-16, rtol_mu=1e-16, tol_sub=1e-8,
+                             armijo_alpha=1e-4, line_search_params=None, stagnation_window=3,
+                             stagnation_threshold=np.inf):
+    """Error aware Trust-Region method for a coercive RB model as surrogate.
+
+    Parameters
+    ----------
+    reductor
+        The reductor used to generate the reduced order models and estimate the output error.
+    primal_dual
+        If `False`, only enrich with the primal solution. If `True`, additionally
+        enrich with the dual solutions.
+
+    See :func:`trust_region`.
+    """
+    if primal_dual:
+        surrogate = PrimalDualTRSurrogate(reductor, initial_guess)
+    else:
+        output = reductor.fom.output_functional
+        if isinstance(output, LincombOperator) and len(output.operators) == 1 and \
+            isinstance(output.operators[0], QuadraticFunctional):
+            # case for quadratic outputs. Should be obsolete soon, see QuadraticOutputTRSurrogate.
+            continuity_estimator_output = output.coefficients[0]
```

Comment:
Doesn't this completely ignore the norm of the actual functional?

Reply:
We now use the `MaxThetaParameterFunctional`.